pkg/scheduler/ugm/queue_tracker.go

```diff
@@ -316,104 +312,132 @@ func (qt *QueueTracker) IsQueuePathTrackedCompletely(queuePath string) bool {
-func (qt *QueueTracker) decreaseTrackedResourceUsageDownwards(queuePath string) map[string]bool {
-	childQueueTracker := qt.getChildQueueTracker(queuePath)
-	childQueueTrackers := childQueueTracker.childQueueTrackers
-	removedApplications := make(map[string]bool)
-	for _, childQT := range childQueueTrackers {
-		if len(childQT.runningApplications) > 0 && childQT.resourceUsage != resources.NewResource() {
-			removedApplications = childQT.runningApplications
-			childQT.resourceUsage = resources.NewResource()
-			childQT.runningApplications = make(map[string]bool)
-			childQT.decreaseTrackedResourceUsageDownwards(childQT.queuePath)
+func (qt *QueueTracker) unlinkAllQT() bool {
+	for childName, childQT := range qt.childQueueTrackers {
+		if childQT.unlinkAllQT() {
+			delete(qt.childQueueTrackers, childName)
 		}
 	}
-	qt.decreaseTrackedResourceUsageUpwards(queuePath)
-	return removedApplications
+	if len(qt.runningApplications) == 0 && len(qt.childQueueTrackers) == 0 {
+		return true
+	}
+	return false
 }
 
-// decreaseTrackedResourceUsageUpwards Traverse upwards all the way upto the root starting from last queue in queuePath,
-// reset resourceUsage and runningApplications to the default value.
-func (qt *QueueTracker) decreaseTrackedResourceUsageUpwards(queuePath string) {
-	childQueuePath, immediateChildQueueName := getChildQueuePath(queuePath)
-	if childQueuePath != common.Empty {
-		if qt.childQueueTrackers[immediateChildQueueName] == nil {
-			log.Log(log.SchedUGM).Error("Child queueTracker tracker must be available in child queues map",
-				zap.String("child queueTracker name", immediateChildQueueName))
+// decreaseTrackedResourceUsageDownwards queuePath either could be parent or leaf queue path.
+// If it is parent queue path, then reset resourceUsage and runningApplications for all child queues,
+// If it is leaf queue path, reset resourceUsage and runningApplications for queue trackers in this queue path.
+func (qt *QueueTracker) decreaseTrackedResourceUsageDownwards(hierarchy []string) map[string]bool {
+	// depth first: all the way to the leaf, ignore if not exists
+	// more than 1 in the slice means we need to recurse down
+	var removedApplications map[string]bool
+	if len(hierarchy) > 1 {
+		childName := hierarchy[1]
+		if qt.childQueueTrackers[childName] != nil {
+			removedApplications = qt.childQueueTrackers[childName].decreaseTrackedResourceUsageDownwards(hierarchy[1:])
 		}
-		qt.childQueueTrackers[immediateChildQueueName].decreaseTrackedResourceUsageUpwards(childQueuePath)
+	} else if len(hierarchy) <= 1 {
+		// reach end of hierarchy, remove all resources under this queue
+		removedApplications = qt.decreaseAllTrackedResourceUsage()
 	}
+
 	if len(qt.runningApplications) > 0 && qt.resourceUsage != resources.NewResource() {
 		qt.resourceUsage = resources.NewResource()
 		qt.runningApplications = make(map[string]bool)
 	}
+
+	return removedApplications
+}
```

Comment:
Can we rewrite this method to avoid newly introduced method `decreaseAllTrackedResourceUsage()` ?

```

func (qt *QueueTracker) decreaseTrackedResourceUsageDownwards(hierarchy []string) map[string]bool {
	log.Log(log.SchedUGM).Debug("decreaseTrackedResourceUsageDownwards ",
		zap.String("current queue ", qt.queueName),
		zap.String("queue path", qt.queuePath),
		zap.Strings("hierarchy", hierarchy),
		zap.Int("no. of child queue trackers", len(qt.childQueueTrackers)))
	// depth first: all the way to the leaf, ignore if not exists
	// more than 1 in the slice means we need to recurse down
	var removedApplications map[string]bool
	if len(hierarchy) > 1 {
		childName := hierarchy[1]
		if qt.childQueueTrackers[childName] != nil {
			removedApplications = qt.childQueueTrackers[childName].decreaseTrackedResourceUsageDownwards(hierarchy[1:])
		}
	} else if len(hierarchy) <= 1 {
		// reach end of hierarchy, remove all resources under this queue
		//removedApplications = qt.decreaseAllTrackedResourceUsage()
		for childName, childQT := range qt.childQueueTrackers {
			if len(childQT.runningApplications) > 0 && childQT.resourceUsage != resources.NewResource() {
				// runningApplications in parent queue should contain all the running applications in child queues,
				// so we don't need to update removedApplications from child queue result.
				childQT.decreaseTrackedResourceUsageDownwards(strings.Split(childName, configs.DOT))
			}
		}
	}

	if len(qt.runningApplications) > 0 && qt.resourceUsage != resources.NewResource() {
		qt.resourceUsage = resources.NewResource()
		qt.runningApplications = make(map[string]bool)
	}
	return removedApplications
}
```

Reply:
Agree. I think this can work. The only different part is that we get child queue **name** from `childQueueTrackers`, not child queue **path**, so we don't need to do `strings.Split` on it.